### ISSUE TITLE
Add State & Country picklist smart support to Contact & Account triggers

### ISSUE THREAD

**Issue by davidhabib** [SalesforceFoundation/NPSP#2067]
When an org has state & country picklists enabled, trying to save a state abbreviation or country abbreviation into the standard state and country fields does not work.  the state & country fields require the full name, and the StateCode & CountryCode fields accept the abbreviations.

Unfortunately, Salesforce does not make StateCode and CountryCode picklists available in custom objects, visualforce, or lightning components.  So in our NPSP Address code, behind the scenes when the user inserts or updates an address, we see if state & country picklists are enabled, and we fixup abbreviations to the fullname automatically.

This work item is to add that same type of fixup functionality in BeforeInsert and BeforeUpdate triggers on Contact and Account.  This will benefit:
1. the Manage Household UI
2. NPSP Data Importer (and other import tools)
3. user's Custom Code writing directly to Contact and Account

**Comment by davidhabib** [comment]
**lurch: add

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x80000000GEw8AAG" target="_blank">W-010235</a>

**Comment by mhiman** [comment]
affects one of our clients, also affecting data imports.

**Comment by mahavrilla** [comment]
At Boston 2016 sprint. Looking at this item.

**Comment by mahavrilla** [comment]
In an attempt to build this enhancement, I have found that there is a known issue when trying to fix the state/country data in a trigger. The state and country validation is enforced prior to before triggers running. 

I executed multiple tests using the API to update a country with CountryCode and have it updated to the label. Each time I would get the following error prior to the before trigger executing.

```
There's a problem with this country, even though it may appear correct. Please select a country/territory from the list of valid countries.: Billing Country
```

Here is the link to the known issue. https://success.salesforce.com/issues_view?id=a1p30000000T4OMAA0
